### ISSUE TITLE
Clearification on Data Structures

### ISSUE THREAD

**Issue by takahser** [aleixmorgadas/nem2-nonfungible-asset#21]
Can you explain how the data is stored on the NEM Catapult ledger?
Which concepts/features do you use? How do you create a new NFT/NFA? How do you make sure, none of the NFT/NFA-related information doesn't get lost while transfering them?

**Comment by takahser** [comment]
Ok so I took a look at the code.

- The [CreateAssetCommand.ts#L50](https://github.com/aleixmorgadas/nem2-nonfungible-asset/blob/master/src/CreateAssetCommand.ts#L50) shows that the NFA is created by sending a message to the NFA address, containing a special payload.

```javascript
    {
        type: CreateAssetCommand.TYPE,
        version: CreateAssetCommand.VERSION,
        source: this.source,
        identifier: this.identifier
     }
``` 

- That NFA address is determined by creating a public key with the NFA identifiers as entropy.

The thing that I didn't understand are:
- how can we prevent others from "creating" an existing NFA?
- how can anybody own a specific NFA?

**Comment by aleixmorgadas** [comment]
Hi @takahser, thanks for sharing an interest in the project :smile: 

TLDR; The ideal scenario is that the NFA be at the protocol level. Until then, we still need a way to represent NFA at the application level and we accept it's a weak implementation compared to protocol implementation.

The architecture of the library is explained on a high level in the [Hypothesis: CQRS Models with Blockchain as persistence layer works](https://medium.com/@aleixmorgadas/hypothesis-cqrs-models-with-blockchain-as-persistence-layer-works-95639acc0cf8) article. It explains the motivations, the pros and cons.

The library isn't at the protocol level, it's at _app level_. It implies that in the way we agreed to read is that determines if an NFA is valid or not. **The library logic is weak compared to protocol logic**, but it's more manageable and faster to develop. Why is it weak? Because we cannot prevent others to write whatever they want. When we read from blockchain, **then** we can discriminate the invalid transactions (based on a logic that we defined).

> That NFA address is determined by creating a public key with the NFA identifiers as entropy.

The public key is `SHA3(source + identifier)`, then I generate the address from that public key. See there's no private key involved on public key generation.

Why source? Different companies might have the same identifier, I needed a way to discriminate between them.

The first message contains the information to validate the address is well generated. When I read the first message, I (or I'll) check if `ADDRESS(SHA3(source+identifier)) == address`. If not, invalid asset representation.

> how can we prevent others from "creating" an existing NFA?

We cannot. The validation at app level happens when we read from the chain. 

It will bring another question to you, what happens if someone changes the logic? 
It's OK for them. The issue will be if you want to interact with others, some apps will read in one way and others in another way, representing the assets in a different way causing a different result.

We think the library should be part of the [NIP](https://github.com/nemtech/NIP/). So, when there's a representation issue, the truth is what the NIP says. Something like `Given state S, When A happens, End state is S'`. In case the result isn't _S'_, we can say that a) there's a bug b) they modified the logic for their own benefit (it's OK as long as they don't pretent to interact with 3rd party).

Does it make sense to you?

> how can anybody own a specific NFA?

1. If it's not created, you create it.
2. If it's created.
2.1. As owner you transfer the ownership to another account. [see issue here](https://github.com/aleixmorgadas/nem2-nonfungible-asset/issues/4)
2.2. As non-owner, you request to the owner the asset transfer (probably, we will use AggregateTransaction here). [see issue here](https://github.com/aleixmorgadas/nem2-nonfungible-asset/issues/6)

-----------------

The lib is under development and we don't know yet if this architecture is viable or not. I think in this library as an experiment for now. I would like to be prudent on using this on production.

I hope it resolves your doubts ^^

**Comment by takahser** [comment]
Hey @aleixmorgadas 

Thanks a lot for your detailed feedback, I appreciate it! 😄
In that case, I think I'll write an ETH ERC 721 token smart contract for now.

> The ideal scenario is that the NFA be at the protocol level.

Yes, that would be great. But it's not on the roadmap so far, is it?

> We think the library should be part of the NIP.

Ah nice, I didn't know there is a project like [BIP](https://github.com/bitcoin/bips) and [EIP](https://github.com/ethereum/EIPs) for NEM already. 👍🏼
Did you get any feedback from the NEM foundation regarding [NIP](https://github.com/nemtech/NIP/blob/master/NIPs/nip-0001.md)?

**Comment by aleixmorgadas** [comment]
> Yes, that would be great. But it's not on the roadmap so far, is it?

No idea.

> Did you get any feedback from the NEM foundation regarding NIP?

Well, NIP isn't from NEM Foundation, it's from NEM itself. I think some NEM Foundation members helps there, but anyone can collaborate.

People shared feedback on the public slack. Check the Slack invitation link at [nemtech](https://nemtech.github.io/), then join there the #nonfungible-asset channel. People shared their opinions there about the library, some contributed to the project with code, others with ideas.

**Comment by takahser** [comment]
Thanks, I joined the slack channel 👍🏼
